service/matching/tasklist/isolation_balancer.go

```diff
@@ -0,0 +1,422 @@
+	for group := range metrics.hasPollersByIsolationGroup {
+		polledOrAssigned[group] = true
+		if _, ok := i.groupState[group]; !ok {
+			i.groupState[group] = i.createIsolationGroupState(false)
+		}
+	}
+	for group := range i.groupState {
+		// No pollers and no write partitions means we don't need to track it
+		if !polledOrAssigned[group] {
+			delete(i.groupState, group)
+			continue
+		}
+	}
+}
+
+func (i *isolationBalancer) createIsolationGroupState(assignToPartitions bool) *isolationGroupState {
+	return &isolationGroupState{
+		underLoad:             clock.NewSustain(i.timeSource, i.config.IsolationGroupDownscaleSustainedDuration),
+		overLoad:              clock.NewSustain(i.timeSource, i.config.IsolationGroupUpscaleSustainedDuration),
+		noPollers:             clock.NewSustain(i.timeSource, i.config.IsolationGroupNoPollersSustainedDuration),
+		hasPollers:            clock.NewSustain(i.timeSource, i.config.IsolationGroupHasPollersSustainedDuration),
+		canAssignToPartitions: assignToPartitions,
+	}
+}
+
+func (i *isolationBalancer) getPartitionsPerGroup(writePartitions map[int]*types.TaskListPartition) map[string]int {
+	partitionCount := make(map[string]int)
+	for _, p := range writePartitions {
+		for _, group := range p.IsolationGroups {
+			partitionCount[group]++
+		}
+	}
+	return partitionCount
+}
+
+func (i *isolationBalancer) calculatePartitionGoal(aggregateMetrics *aggregatePartitionMetrics, partitionCountByGroup map[string]int, writePartitionCount int) map[string]int {
+	partitionGoal := make(map[string]int)
+	assignableGroups := 0
+	// Identify groups that have switched to assignable and count them. This impacts the minimum partitions per group
+	for group, state := range i.groupState {
+		groupHasPollers := aggregateMetrics.hasPollersByIsolationGroup[group]
+		sustainedPollers := state.hasPollers.CheckAndReset(groupHasPollers)
+		sustainedNoPollers := state.noPollers.CheckAndReset(!groupHasPollers)
+		if state.canAssignToPartitions {
+			if sustainedNoPollers {
+				i.scope.Tagged(metrics.IsolationGroupTag(group)).IncCounter(metrics.IsolationGroupStoppedPolling)
+				state.canAssignToPartitions = false
+			} else {
+				assignableGroups++
+			}
+		} else if sustainedPollers {
+			i.scope.Tagged(metrics.IsolationGroupTag(group)).IncCounter(metrics.IsolationGroupStartedPolling)
+			state.canAssignToPartitions = true
+			state.overLoad.Reset()
+			state.underLoad.Reset()
+			assignableGroups++
+		}
+	}
+
+	minPartitions := i.getMinPartitionsPerGroup(writePartitionCount, assignableGroups)
+
+	for group, state := range i.groupState {
+		// Remove all assignments from inactive groups
+		if !state.canAssignToPartitions {
+			partitionGoal[group] = 0
+			i.scope.Tagged(metrics.IsolationGroupTag(group)).UpdateGauge(metrics.IsolationGroupPartitionsGauge, 0)
+			continue
+		}
+		currentPartitions := partitionCountByGroup[group]
+		groupQPS := aggregateMetrics.qpsByIsolationGroup[group]
```

Comment:
Question for my understanding: 
This is getting the RPS value per tasklist total from the rate-limiter? Is this defined in all cases?

Reply:
oh, I must have missed that, sorry, misread.